File: packages/health_connector_hc_android/android/src/main/kotlin/com/phamtunglam/health_connector_hc_android/mappers/health_record_mappers/RestingHeartRateRecordMapper.kt
package com.phamtunglam.health_connector_hc_android.mappers.health_record_mappers

import androidx.health.connect.client.records.RestingHeartRateRecord
import com.phamtunglam.health_connector_hc_android.mappers.metadata_mappers.toDto
import com.phamtunglam.health_connector_hc_android.mappers.metadata_mappers.toHealthConnect
import com.phamtunglam.health_connector_hc_android.pigeon.RestingHeartRateRecordDto
import java.time.Instant
import java.time.ZoneOffset

/**
 * Converts a Health Connect [RestingHeartRateRecord] to a [RestingHeartRateRecordDto].
 */
internal fun RestingHeartRateRecord.toDto(): RestingHeartRateRecordDto = RestingHeartRateRecordDto(
    id = metadata.id,
    time = time.toEpochMilli(),
    zoneOffsetSeconds = zoneOffset?.totalSeconds?.toLong(),
    metadata = metadata.toDto(),
    beatsPerMinute = beatsPerMinute.toDouble(),
)

/**
 * Converts a [RestingHeartRateRecordDto] to a Health Connect [RestingHeartRateRecord].
 */
internal fun RestingHeartRateRecordDto.toHealthConnect(): RestingHeartRateRecord =
    RestingHeartRateRecord(
        beatsPerMinute = beatsPerMinute.toLong(),
        time = Instant.ofEpochMilli(time),
        zoneOffset = zoneOffsetSeconds?.let { ZoneOffset.ofTotalSeconds(it.toInt()) },
        metadata = metadata.toHealthConnect(id),
    )

File: packages/health_connector_hc_android/lib/src/mappers/health_record_mappers/heart_rate/resting_heart_rate_record_mapper.dart
import 'package:health_connector_core/health_connector_core_internal.dart'
    show HealthRecordId, RestingHeartRateRecord, sinceV1_3_0, Frequency;
import 'package:health_connector_hc_android/src/mappers/health_record_mappers/health_record_id_mapper.dart';
import 'package:health_connector_hc_android/src/mappers/metadata_mappers/metadata_mapper.dart';
import 'package:health_connector_hc_android/src/pigeon/health_connector_hc_android_api.g.dart'
    show RestingHeartRateRecordDto;
import 'package:meta/meta.dart' show internal;

/// Converts [RestingHeartRateRecordDto] to [RestingHeartRateRecord].
@sinceV1_3_0
@internal
extension RestingHeartRateRecordDtoToDomain on RestingHeartRateRecordDto {
  RestingHeartRateRecord toDomain() {
    return RestingHeartRateRecord.internal(
      id: id?.toDomain() ?? HealthRecordId.none,
      time: DateTime.fromMillisecondsSinceEpoch(time, isUtc: true),
      zoneOffsetSeconds: zoneOffsetSeconds,
      metadata: metadata.toDomain(),
      rate: Frequency.perMinute(beatsPerMinute),
    );
  }
}

/// Converts [RestingHeartRateRecord] to [RestingHeartRateRecordDto].
@sinceV1_3_0
@internal
extension RestingHeartRateRecordToDto on RestingHeartRateRecord {
  RestingHeartRateRecordDto toDto() {
    return RestingHeartRateRecordDto(
      id: id.toDto(),
      time: time.millisecondsSinceEpoch,
      zoneOffsetSeconds: zoneOffsetSeconds,
      metadata: metadata.toDto(),
      beatsPerMinute: rate.inPerMinute,
    );
  }
}
